harness/determined/cli/master.py

```diff
@@ -22,6 +24,17 @@ def config(args: Namespace) -> None:
         raise ValueError(f"Bad output format: {args.output}")
 
 
+def get_master(args: Namespace) -> None:
+    session = setup_session(args)
+    resp = get_GetMaster(session._do_request)
```

Comment:
passing session itself may look more natural (boto at the call site
```suggestion
    session = setup_session(args)
    resp = get_GetMaster(session)
```

a class instance method style like `session.get_GetMaster` would probably be even more in line with things like boto3.

Reply:
The idea here was to leave the (experimental) determined package out of the generated code. This way you can pass in any request method that supports the same parameters and use the API instead of having to have determined installed.
Although if we're strictly using it as the examples here show `session` works as well

Reply:
maybe we could use something like that to have typing for the `session` parameter while avoiding imports: https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols

Reply:
>This way you can pass in any request method that supports the same parameters and use the API instead of having to have determined installed.

that makes sense as a feature; however passing around an underscored property `session._do_request`  seems too weird for me in our code.

Reply:
> maybe we could use something like that to have typing for the session parameter while avoiding imports: https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols

it'd still mean that the user of the bindings would have to pass in a class that conforms to what a determined `session` looks like. how does this look? 🤔

Reply:
I don't think we're here to write a swagger bindings generator that works for anybody but us.  That will keep this code simple and keep its surface area small.  So requiring determined to be installed makes perfect sense.

Reply:
hmm true but if there is no real cost to allowing the generated bindings to be used without the determined package being installed why not do it? It'd still allow people interested in working with determined to use the bindings without depending on another package. Does it look off without passing the session? Is there a way I could set it up so that it doesn't look off?

Reply:
> a class instance method style like session.get_GetMaster would probably be even more in line with things like boto3.

a class instance would be a different style and we can certainly consider that as well.

Reply:
> a class instance method style like session.get_GetMaster would probably be even more in line with things like boto3

I don't think "being like boto3" is actually desirable.  That library (and the botocore library behind it) accomplish very little in an absolutely enormous volume of code.  Subclassing parts of it to get custom behavior is very difficult.  See our boto3_credential_manager.py for example.> It'd still allow people interested in working with determined to use the bindings without depending on another package.

No, the swagger spec itself is what you give to these people.  If they want to use Determined without using the python package, then they should generate their own bindings from the swagger spec.  That's like the stated goal of publishing the swagger spec.

> how does this look

Also, where are these people?  I've never heard of a single user trying to use the REST API directly.  I've heard of a lot of people who want to see good python bindings.

The introduction of `client.http_request` means that in practice the bindings still depend on a Session, only instead of being an explicit Session as before, now it's an implicit singleton Session.  Also, it's the wrong Session; there are already singletons for the cli's master certs and session tokens in the code, but using the user-facing Session embedded in the client module will cause bugs.

> hmm true but if there is no real cost to allowing the generated bindings to be used without the determined package being installed why not do it?

I would argue there is a real cost to what you have.  It's another layer of indirection in how our own code works.  The Session object wraps the requests library in the cert handling and token handling that we need.  Having another function that wraps the Session object and has an API kinda like the requests library is another layer of indirection.  Our developers have to read and understand what's going on, so extra layers of indirection have a very real cost, especially for onboarding new engineers.

In 


I think it's important to remember that we are adding swagger bindings for the type-checking behavior that they give us.  They are not particularly suitable for giving to a user (a function named `get_GetMaster()`... that would never pass code review for a user-api) because they're just clunky.  Everywhere that we want to have a user-facing Python API  
- We want type checking

And why we are not adding swagger bindings:
-

Reply:
> It'd still allow people interested in working with determined to use the bindings without depending on another package.

No, the swagger spec itself is what you give to these people.  If they want to use Determined without using the python package, then they should generate their own bindings from the swagger spec.  That's like the stated goal of publishing the swagger spec.

Also, where are these people?  I've never heard of a single user trying to use the REST API directly.  I've heard of a lot of people who want to see good python bindings.

> how does this look

The introduction of `client.http_request` means that in practice the bindings still depend on a Session, only instead of being an explicit Session as before, now it's an implicit singleton Session.  This is worse because it is harder to understand.  Also, it's the wrong Session; there are already singletons for the cli's master certs and session tokens in the code, but using the user-facing Session embedded in the client module will cause bugs.

> hmm true but if there is no real cost to allowing the generated bindings to be used without the determined package being installed why not do it?

I would argue there is a real cost to what you have.  It's another layer of indirection in how our own code works.  The Session object wraps the requests library in the cert handling and token handling that we need.  Having another function that wraps the Session object and has an API kinda like the requests library is another layer of indirection.  Our developers have to read and understand what's going on, so extra layers of indirection have a very real cost, especially for onboarding new engineers.

Ultimately, I'm arguing for the absolute simplest solution that gives us type-checking.  Early generalization is a form of early optimization, and I think it just makes our lives harder without making any user lives' easier.

Reply:
> a class instance would be a different style and we can certainly consider that as well.

This I could be ok with, but if you do go this route please make that giant generated class separate from the Session itself.  Make it some SessionContainingObject that takes a Session as input (i.e. separate the generated code from the hand-written code).

Personally, I still think that all of these generated APIs are always clunky.  You can never give them to end users directly because names like `get_GetMaster()` are absolutely terrible names.  If you hand-wrote a function like that and tried to put it in a user-facing API it would never, never make it through code-review.  Therefore, since these generated APIs are too clunky to give to users, we'll always end up writing wrappers around them.  This is basically what the Deterimined() object is; a bunch of user-friendly wrappers around requests against the REST APIs.  So when we write those wrappers, wrapping a function that takes a Session vs wrapping an object method that was already given a Session seems basically equally clunky from the determined developer's perspective, and I wouldn't bother.

Reply:
I'd say the generator itself is added value if anyone wanted to use the rest apis in python. As opposed to asking them to generate their own since if there already was a way of generating the bindings that's as fast as we wanted it we wouldn't be writing our own generator. Now if we don't consider that use case at all then the point that `client.http_request` would make our code somewhat harder to read IMO is enough to have it take a `determined session` instead.